zephyr/CMakeLists.txt

```diff
@@ -43,52 +106,73 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
     --pad-header
     )
 
-  add_custom_command(
-    OUTPUT
-    ${signed_image_hex}
-    ${update_hex}
-    ${update_bin}
-    COMMAND
-    ${sign_cmd}
-    ${to_sign_hex}
-    ${signed_image_hex}
-    COMMAND
-    ${CMAKE_OBJCOPY}
-    --input-target=ihex
-    --output-target=binary
-    ${to_sign_hex}
-    ${to_sign_bin}
-    COMMAND
-    ${sign_cmd}
-    ${to_sign_bin}
-    ${update_bin}
-    COMMAND
-    ${sign_cmd}
-    --pad # This argument is needed for MCUboot to apply the test swap.
-    ${to_sign_hex}
-    ${test_update_hex}
-    COMMAND
-    ${sign_cmd}
-    ${to_sign_hex}
-    ${update_hex}
-    COMMAND
-    ${CMAKE_OBJCOPY}
-    --input-target=ihex
-    --output-target=ihex
-    --change-address $<TARGET_PROPERTY:partition_manager,PM_MCUBOOT_PRIMARY_SIZE>
-    ${update_hex}
-    ${PROJECT_BINARY_DIR}/moved_update.hex
-    DEPENDS
-    ${sign_depends}
+  set(app_offset $<TARGET_PROPERTY:partition_manager,PM_MCUBOOT_PRIMARY_SIZE>)
+
+  sign(${app_to_sign_hex}     # Hex to sign
+    ${PROJECT_BINARY_DIR}/app # Prefix for generated files
+    ${app_offset}             # Offset
+    ${app_sign_depends}       # Dependencies
+    app_signed_hex            # Generated hex output variable
     )
-  add_custom_target(mcuboot_sign_target DEPENDS ${signed_image_hex})
+
+  add_custom_target(mcuboot_sign_target DEPENDS ${app_signed_hex})
 
   set_property(GLOBAL PROPERTY
     mcuboot_primary_app_PM_HEX_FILE
-    ${signed_image_hex}
+    ${app_signed_hex}
     )
   set_property(GLOBAL PROPERTY
     mcuboot_primary_app_PM_TARGET
     mcuboot_sign_target
     )
+
+  if (CONFIG_MCUBOOT_BUILD_S1_VARIANT)
+    # Secure Boot (B0) is enabled, and we have to build update candidates
+    # for both S1 and S0.
+
+    foreach(slot s0_image;s1_image)
+      string(SUBSTRING ${slot} 0 2 slot_only) # 's0_image' -> 's0'
+
+      # Fetch the target and hex file for the current slot.
+      # Note that these hex files are already signed by B0.
+      get_property(${slot}_target GLOBAL PROPERTY ${slot}_PM_TARGET)
+      get_property(${slot}_hex GLOBAL PROPERTY ${slot}_PM_HEX_FILE)
+
+      # The gap from S0/S1 partition is calculated by partition manager
+      # and stored in its target.
+      set(slot_offset
+        $<TARGET_PROPERTY:partition_manager,${slot_only}_TO_SECONDARY>)
+
+      # Depend on both the target for the hex file, and the hex file itself.
+      set(dependencies "${${slot}_target};${${slot}_hex}")
+
+      set(out_path ${PROJECT_BINARY_DIR}/signed_by_mcuboot_and_b0_${slot})
+
+      sign(${${slot}_hex} # Hex file to sign
+        ${out_path}
+        ${slot_offset}
+        "${dependencies}"    # Need "..." to make it a list.
+        b_one_signed_hex     # Created file variable
```

Comment:
b_one == b1? IMO "signed_hex" is sufficient.

Reply:
fixed